### DISCUSSION TITLE
5 seconds time delay after websend between Tasmota device and computer

### DISCUSSION THREAD

**Discussion by TPPTim** [arendst/Tasmota#17793] [Q&A] [answered]
Hey folks,
i got an time problem with the command websend.
First of all, i try to use websend in rules to send self defined strings to my self developed smart home system (short: SHS) via TCP (i don´t use MQTT).

In my first tests, i received the message from the Tasmota device in my SHS but got in Tasmota console an **Websend: connect failed** back. Also there was an delay of 5 seconds between triggering the rule and got back the result **Websend: connect failed**.

I figured out, that Tasmota expected an **HTTP/1.1 200 OK** back, to recognized if the message was received. After that, i implemented this feedback in my SHS. Now, after i receive the message, i send the receipt confirmation back. This changed the answer in Tasmota console to **Websend: Done** after triggering the rule.
See this picture:
![Unbenannt1](https://user-images.githubusercontent.com/95652405/214784144-94c91e0f-f5c7-497a-8c3a-20f8b9fb9e95.JPG)

If i use the same rule on the same device to send an command like **POWER1 ON** to another Tasmota device it works near instand.
See this picture:
![Unbenannt3](https://user-images.githubusercontent.com/95652405/214789035-e22ac253-7ec7-4e75-b6db-92c215741e57.JPG)

These are the zwo rules i used for testing:
![Unbenannt2](https://user-images.githubusercontent.com/95652405/214789548-9e79be99-a9f3-41d2-b922-7fad46457caa.JPG)

So there must be an difference between the feedback from the Tasmota device and my computer that received the TCP package and send back the receipt confirmation. Because i think that the 5 seconds are a kind of timeout.
This is the string i send back which changed the Websend state from **connect failed** to **Done**:
`HTTP/1.1 200 OK /r/nConnection: close /r/n/r/n`

Can someone help me with this problem? What does the Tasmota device needs to receive that it will not get into 5 seconds timeout?

Best regards
Tim

**Comment by barbudor** [comment]
Try to get better logs like using wireshark/tcpdump
It's more likely your pc software is slow to answer

**Reply by sfromis** [reply]
Doing a simple test with a local web server, I saw less than 0.1 seconds latency from the WebSend command to getting the "Done" from the status 200 ok response, matching the web server being quick to respond. The web server (coded by me) is a very simple one which is just working normally, sending the response and closing the connection.

**Reply by TPPTim** [reply]
Thanks for your answers!
I will do a traffic log on my fritzbox and will analyze it via wireshark. Probably i can see the needed information.
@sfromis Thanks for testing the described behavior. Your result is really interesting, because i also coded an simple and minimal webserver.
It waits for connections on port 80. If someone connected, i read out the message and replied with the respond string. After that i close the connection. So for me it seams that we do the same sequence.

Can you post or share your minimal webserver project/file?

**Reply by sfromis** [reply]
My web server is not a separate project/file, but a feature integrated into the mechanics of a unique programming framework I created for myself, which is not really suitable for release. It is based on server sockets in Java, meaning that I avoided the deeper technical details of managing TCP traffic.

When responding to a http request, I found it preferable to include the Date, Content-Length and Content-Type headers, but I don't remember what issues caused me to add those fields; it was not "just to be nice".

**Reply by TPPTim** [reply]
Thanks @sfromis for your explanation. I will try to change the response. Lets see what happened. I will also do a detailed timestamp protocol to realize at which point the delay will be create. I´ll come back.

**Reply by sfromis** [reply]
BTW... Part of the http protocol is to remember an empty line between the headers and the body.

**Comment by TPPTim** [comment, marked as answer]
Thanks for your input @sfromis and @barbudor !
Yesterday I fixed the 5 second problem. This is the solution. In advance i can say, its not a Tasmota related problem.

I did an exact log of doing websend on Tasmota device and also in my simple webserver application.
I figured out, that the 5 second time delay was not between receiving the message and sending the acknowledge located.
The TCP lister function i used to wait for an incoming connection used an default value for an input parameter that i never used before. Its called resolve network address. The function asks, before executing and returning me the listener reference, the DNS Server to resolve the incoming IP address to the used clear name of the device. And exact this communication with the DNS Server i don´t need has taken 5 seconds. After disable this option via the right parameter, it works like a charm.

The delay between **Websend** and receiving the **Websend Done** answer is <50ms!

Best Regards
Tim
